File: smalivm/src/main/java/org/cf/smalivm/opcode/SwitchPayloadOpFactory.java
package org.cf.smalivm.opcode;

import gnu.trove.map.TIntIntMap;
import gnu.trove.map.TIntObjectMap;
import gnu.trove.map.hash.TIntIntHashMap;

import org.cf.smalivm.VirtualMachine;
import org.jf.dexlib2.builder.MethodLocation;
import org.jf.dexlib2.iface.instruction.SwitchElement;
import org.jf.dexlib2.iface.instruction.SwitchPayload;

public class SwitchPayloadOpFactory implements OpFactory {

    @Override
    public Op create(MethodLocation location, TIntObjectMap<MethodLocation> addressToLocation, VirtualMachine vm) {
        SwitchPayload instr = (SwitchPayload) location.getInstruction();
        TIntIntMap targetKeyToOffset = new TIntIntHashMap();
        for (SwitchElement element : instr.getSwitchElements()) {
            targetKeyToOffset.put(element.getKey(), element.getOffset());
        }

        return new SwitchPayloadOp(location, addressToLocation, targetKeyToOffset);
    }

}

File: smalivm/src/main/java/org/cf/smalivm/opcode/SwitchPayloadOpFactory.kt
package org.cf.smalivm.opcode

import org.cf.smalivm.VirtualMachine
import org.jf.dexlib2.builder.MethodLocation
import org.jf.dexlib2.iface.instruction.SwitchPayload

class SwitchPayloadOpFactory : OpFactory {
    override fun create(location: MethodLocation, addressToLocation: Map<Int, MethodLocation>, vm: VirtualMachine): Op {
        val instr = location.instruction as SwitchPayload
        val targetKeyToOffset: HashMap<Int, Int> = HashMap()
        for (element in instr.switchElements) {
            targetKeyToOffset[element.key] = element.offset
        }
        return SwitchPayloadOp(location, addressToLocation, targetKeyToOffset)
    }
}
